### PR TITLE
alternator: implement ProjectionType=KEYS_ONLY for GSI and LSI

### PR DESCRIPTION
**Pull request by nyh**
In this patch we add support in Alternator's GSI and LSI for the ProjectionType=KEYS_ONLY option, which allows the index to not copy all not _all_ the attributes from the base table to the materialized view - as ProjectionType=ALL does and was the only option supported today. Before this patch, if ProjectionType=KEYS_ONLY was used, it was silently ignored and wasted space - the GSI/LSI was much bigger than it was supposed to be.

Note that the last remaining ProjectionType option, "INCLUDE", is still *not* supported in this patch, which is why this patch Refs #5036 but doesn't yet "Fixes" it. To implement KEYS_ONLY, as we do in this patch, we need to fully drop the ":attrs" map (the map holding all the non- key attributes) from the materialized view. Implementing INCLUDE will require more elaborate code which modifies a map - keeping only specific elements - while copying it from the base-table to view, a feature that we don't yet support in Scylla's materialized view implementation.

    Note that Alternator's behavior regarding ProjectionType=INCLUDE changes
    after this patch: Before this patch it was silently (and incorrectly)
    ignored - the new GSI/LSI contained all attributes instead of just a
    select few. So the application could read the attributes it cared
    about - but also other attributes. After this patch, INCLUDE is just
    rejected as an error. If a user wants to fallback to using ALL, they
    will need to change their application to use ALL.

Changes:

* CreateTable / UpdateTable: parse the Projection parameter for LSI and GSI definitions. For ProjectionType=KEYS_ONLY, omit the ":attrs" column from the materialized view schema. For ALL (the default), the view includes ":attrs" as before. ProjectionType=INCLUDE is rejected as not yet supported. NonKeyAttributes is rejected when ProjectionType is not INCLUDE.

* DescribeTable: report the correct ProjectionType (KEYS_ONLY or ALL) for each LSI/GSI by inspecting whether the view schema contains the ":attrs" column (via the new projection_is_keys_only() helper).

* Query on GSI with KEYS_ONLY: reject Select=ALL_ATTRIBUTES with a clear error, matching DynamoDB behavior (a GSI with KEYS_ONLY cannot do base-table lookups).

* Query on LSI with KEYS_ONLY: unlike GSI, DynamoDB allows Select=ALL_ATTRIBUTES on a KEYS_ONLY LSI. as the existing test test_lsi.py::test_lsi_get_not_projected_attribute shows, after reading the key columns from the LSI, we need do read the unprojected columns from the base table - this is similar to CQL's secondary index where a query begins by finding only the list of keys in the view but then needs to go to the base table to read each of these matching rows individually. Note that the second step is fairly *inefficient*: Despite the fact that in an LSI the base-table partition is on the same replica as the LSI's partition, they are not ordered the same so the second step is an inefficient loop of many random-access single-row reads - not a scan. We implement this part in a new do_query_lsi_with_base_fetch() function that:
    1. Queries the LSI view to get results in LSI sort-key order.
    2. For each row, extracts the base-table primary key from the LSI result (the LSI view schema always includes both the LSI sort key and the base clustering key).
    3. Does a point read on the base table to retrieve the full item. This preserves LSI sort order while returning complete item data.

* Added a metric scylla_alternator_lsi_reads_from_base_table which counts the number of inefficient base-table reads (count of rows) needed for LSI reads that ask to read non-projected attributes. This metric can be useful for users (who will want to avoid these inefficient reads as much as possible), but also important for our test that verifies that these inefficient base-table reads do not happen when not necessary.

* Tests: Remove now-passing xfail markers from test_gsi.py and test_lsi.py, and also add new tests for additional corner cases discovered during this development. As usual, all tests pass also on DynamoDB.

### CODE REVIEW

**Comment by nyh** [comment]
I've left this pull-request is draft because although it passes all the existing xfail tests for `ProjectionType=KEYS_ONLY` (and doesn't break any existing passing tests, most notably `test_lsi.py`::`test_lsi_get_not_projected_attribute`), there are still a bunch more tests I want to write to confirm things that I didn't think of while writing the original selection of xfailing tests for `ProjectionType=KEYS_ONLY`.

In any case, as the commit message says, this patch does not fully "Fixes" https://github.com/scylladb/scylladb/issues/5036, because it only implements `ProjectionType=KEYS_ONLY` (and of course `ProjectionType=ALL`), not `ProjectionType=INCLUDE`.

**Comment by doublex** [comment]
Great feature!

**Comment by nyh** [comment]
Pushed new version: continued to write more tests and found one bug and fixed it.

I still have a list of more tests I want to write for corner cases which I believe that the current tests are not covering, and might still have bugs.

**Comment by nyh** [comment]
Pushed a new version with a few more bugs fixed, and many more tests.
I think this version is good enough to go out of "draft" mode.

**Comment by nyh** [comment]
@scylladb/releng I think something is broken with the bots, I get a failure on "PR require backport label" but it does have a backport/none label.

**Comment by yaronkaikov** [comment]
> @scylladb/releng I think something is broken with the bots, I get a failure on "PR require backport label" but it does have a backport/none label.

Just a random failure with the github action, i re-run it and it passed

File: alternator/executor_read.cc [hunk 1]
```diff
@@ -609,6 +647,151 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
...
+                base_schema->id(), base_schema->version(), base_partition_slice,
+                proxy.get_max_result_size(base_partition_slice),
+                query::tombstone_limit(proxy.get_tombstone_limit()));
+        service::storage_proxy::coordinator_query_result qr =
+                co_await proxy.query(base_schema, base_command,
+                        {dht::partition_range(dht::decorate_key(*base_schema, base_pk))},
+                        cl,
+                        service::storage_proxy::coordinator_query_options(
+                                timeout, permit, client_state, trace_state));
+        ++stats.lsi_reads_from_base_table;
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor_read.cc:793]: Very good catch. If we do have per-table metrics (it's a separate question whether they are always needed, but we have them), they need to be updated.

File: test/alternator/test_gsi.py [hunk 2]
```diff
@@ -2413,3 +2402,255 @@ def test_gsi_query_exclusivestartkey_spurious_column(test_table_gsi_5):
...
+# Also check that if the Projection option is missing, the default is ALL
+# and should be reported as such by DescribeTable.
+# We test creating the GSI to be described in two ways - via CreateTable
+# (creating a base table and a GSI together) and via UpdateTable (adding a
+# new GSI to an existing table).
+@pytest.mark.parametrize('projection_type', [
+    None,
+    'ALL',
+    'KEYS_ONLY',
+    pytest.param('INCLUDE', marks=pytest.mark.xfail(reason='ProjectionType=Issue #5036 - ProjectionType=INCLUDE not yet supported')),
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_gsi.py]: Very true.

File: test/alternator/test_lsi.py [hunk 3]
```diff
@@ -386,9 +387,179 @@ def test_lsi_get_not_projected_attribute(test_table_lsi_keys_only):
                        'b': {'AttributeValueList': [b2], 'ComparisonOperator': 'EQ'}},
         Select='SPECIFIC_ATTRIBUTES', AttributesToGet=['d'])
 
+# Check that querying an entire LSI partition (all items with the same hash
+# key) works correctly: ALL_PROJECTED_ATTRIBUTES returns only the projected
+# attributes (in this case, key columns p, c, b), while ALL_ATTRIBUTES returns
+# the full items (including the non-projected attribute d). The items are
+# returned sorted by the LSI sort key 'b', which is a different order than the
+# base table sort key 'c'.
+def test_lsi_get_whole_partition(test_table_lsi_keys_only):
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_lsi.py:396]: Very true. Good catch.

File: test/alternator/test_lsi.py [hunk 4]
```diff
@@ -386,9 +387,179 @@ def test_lsi_get_not_projected_attribute(test_table_lsi_keys_only):
...
+    # (not even the key columns are returned).
+    assert all(set(r.keys()) == {'d'} for r in results)
+    # Query with Select=ALL_ATTRIBUTES: also does one base-table read per item.
+    before = get_metric(metrics, metric)
+    results = full_query(test_table_lsi_keys_only, IndexName='hello',
+        KeyConditions={'p': {'AttributeValueList': [p], 'ComparisonOperator': 'EQ'}},
+        Select='ALL_ATTRIBUTES')
+    after = get_metric(metrics, metric)
+    assert len(results) == len(items)
+    assert after - before == len(items)
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_lsi.py:663]: We actually have a way to solve this, by using per-table metrics! I'll switch to it. Thanks for nudging me to stop ignoring this problem.

**Comment by nyh** [comment]
Pushed a new version fixing a bunch of good suggestions by copilot.

**Comment by nyh** [comment]
Amazing, it seems Copilot's review quality gets better every day. It now found bugs it didn't find yesterday - not to mention that I missed them too.

**Comment by nyh** [comment]
@scylladb/releng the CI doesn't seem to have started after 14 hours.

**Comment by yaronkaikov** [comment]
> @scylladb/releng the CI doesn't seem to have started after 14 hours.

Looks like a bug, checking it

File: alternator/executor_read.cc [hunk 5]
```diff
@@ -609,6 +647,152 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
...
+        old_paging_state = make_lw_shared<service::pager::paging_state>(pk, pos, query::max_partitions, query_id::create_null_id(), service::pager::paging_state::replicas_per_token_range{}, std::nullopt, 0);
+    }
+
+    co_await verify_permission(enforce_authorization, warn_authorization, client_state, base_schema, auth::permission::SELECT, stats);
+
+    // Step 1: Query the LSI to get key-only rows in LSI sort order.
+    // The LSI is KEYS_ONLY so it has no ":attrs" column, but it does have
+    // all DynamoDB key columns: the base partition key (p), the LSI sort key (b),
+    // and the base clustering key (c). We need p and c to look up the base table.
+    auto lsi_regular_columns =
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor_read.cc:707]: Excellent catch! I wrote a new test in test_cql_rbac.py to catch this mistake, and fixed it.

**Comment by yaronkaikov** [comment]
> > @scylladb/releng the CI doesn't seem to have started after 14 hours.
> 
> Looks like a bug, checking it

@nyh i re-run the CI now it's ok. The issue was some temp slow response from Jenkins API, sent a patch in increase it - https://github.com/scylladb/scylla-pkg/pull/6122

File: alternator/executor_read.cc [hunk 6]
```diff
@@ -609,6 +647,152 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
...
+    }
+
+    rjson::value response = rjson::empty_object();
+    rjson::add(response, "Count", rjson::value(matched_count));
+    rjson::add(response, "ScannedCount", rjson::value(scanned_count));
+    rjson::add(response, "Items", std::move(items_json));
+    if (paging_state) {
+        rjson::add(response, "LastEvaluatedKey", encode_paging_state(*lsi_schema, *paging_state));
+    }
+    co_return rjson::print(std::move(response));
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor_read.cc]: Good catch.

File: alternator/executor_read.cc [hunk 7]
```diff
@@ -780,11 +964,29 @@ future<executor::request_return_type> executor::scan(client_state& client_state,
...
+            "Select=ALL_ATTRIBUTES is not supported for GSI with ProjectionType=KEYS_ONLY. Use Select=ALL_PROJECTED_ATTRIBUTES or a ProjectionExpression instead."));
+    }
 
     std::unordered_set<std::string> used_attribute_names;
     std::unordered_set<std::string> used_attribute_values;
     auto attrs_to_get = calculate_attrs_to_get(request, *_parsed_expression_cache, used_attribute_names, select);
 
+    // For GSI with KEYS_ONLY, validate that specific requested attributes
+    // are key columns (projected into the index).
+    if (attrs_to_get && table_type == table_or_view_type::gsi && projection_is_keys_only(*schema)) {
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor_read.cc:1014]: Wow, excellent catch. Both the AI that wrote this code, and I that steered it and reviewed it, missed this bug. I'll write a test for this and fix it.

**Comment by nyh** [comment]
Sent a version fixing the new bugs that copilot found. So AI writes the code, another AI finds its bugs, the first AI fixes these bugs... I think I can retire now ;-)

**Comment by nyh** [comment]
This is a feature we've been missing for years (since #5036) so we can go with it missing for another release. No backport to 2026.2.

File: alternator/executor_read.cc [hunk 8]
```diff
@@ -609,6 +647,163 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
...
+                        cl,
+                        service::storage_proxy::coordinator_query_options(
+                                timeout, permit, client_state, trace_state));
+        ++stats.lsi_reads_from_base_table;
+        ++get_stats_from_schema(proxy, *base_schema)->lsi_reads_from_base_table;
+        auto opt_item = describe_single_item(base_schema, base_partition_slice,
+                *base_selection, *qr.query_result, *shared_attrs_to_get);
+        if (opt_item && (!filter || filter.check(*opt_item))) {
+            ++matched_count;
+            items.push_back(std::move(*opt_item));
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor_read.cc:799]: Actually, it turns out that this is not a problem: DynamoDB forbids filtering on non-projected columns, **even if** this non-projected column is selected. So the request described in your scenario will be an error. You're right, it didn't have to be, and DynamoDB (and Alternator) could have supported it - but it does not. We even have a test for this.

File: test/alternator/test_lsi.py [hunk 9]
```diff
@@ -386,9 +387,242 @@ def test_lsi_get_not_projected_attribute(test_table_lsi_keys_only):
...
+                ExpressionAttributeValues={':val': '2'})
+    # Curiously, there remains one single attribute we may filter on - c.
+    # This is because c is not officially part of the LSI key schema, but
+    # is still projected because it is one of the base table's key attributes
+    # so DynamoDB does allow filtering on it.
+    results = full_query(test_table_lsi_keys_only, IndexName='hello',
+        KeyConditions={'p': {'AttributeValueList': [p], 'ComparisonOperator': 'EQ'}},
+        FilterExpression='c = :val',
+        ExpressionAttributeValues={':val': '2'})
+    assert len(results) == 1 and results[0]['c'] == '2'
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_lsi.py:526]: Good catch, this actually caught a bug. It means your previous comment was more correct than I thought - even though in general it's not allowed to filter on a non-projected column so  what you said earlier about the missing feature was not necessary - it's necessary in this specific weird corner case of 'c'.

File: alternator/executor_read.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor_read.cc:799]: Actually you were right and this turned out to be a problem, in some rare situations (like selecting a non-key column but filtering on a key column) - but it's a problem. I'm adding a test and a fix.

**Comment by nyh** [comment]
Pushed another version after fixing another batch of bugs discovered by copilot. That's amazing - it found very unlikely and very rare bugs - but at the same time - very real bugs that are very easy to reproduce in tests (I did), and actually quite easy to fix.

File: alternator/executor.cc [hunk 10]
```diff
@@ -1649,15 +1692,19 @@ future<executor::request_return_type> executor::create_table_on_shard0(service::
...
+            for (const column_definition& regular_cdef : schema->regular_columns()) {
+                if (!view_builder.has_column(*cql3::to_identifier(regular_cdef))) {
+                    view_builder.with_column(regular_cdef.name(), regular_cdef.type, column_kind::regular_column);
+                }
             }
         }
+        // NOTE: as we explained in an earlier comment, when view_keys_only
+        // and we don't copy some columns to the view (notably ":attrs"),
+        // we still don't need to add virtual columns to the view.
         const bool include_all_columns = true;
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor.cc]: Looking deeper, I think this comment is not correct, but it's conclusion is:
The materialized view code currently always reads the entire base-table row, the "include_all_columns" flag doesn't change it.

However, as far as I understand, the main raison d'etre for this flag is that in CQL's DESCRIBE MATERIALIZED VIEW for a view selecting everything it will show "`SELECT *`" instead of SELECT of all the specific comments - and also ALTER TABLE adding a column will know if to add the new column (`SELECT *` case) or not (SELECT of specific columns case). None of this is critical for Alternator - but it will be better that DESCRIBE MATERIALIZED VIEW will not show "SELECT *" in this case, so I'll fix it.

**Comment by nyh** [comment]
Fixed another suggestion from copilot - this one getting very minor (how a KEYS_ONLY GSI/LSI looks to CQL's "DESCRIBE TABLE") but was worth doing anyway.

**Comment by nyh** [comment]
I just realized there's something missing in this pull request - fixing RBAC (authorization):

As we test in `test_cql_rbac.py`::`test_rbac_query_separate_index_permissions`, a base table and its GSIs/LSIs have separate permissions - you can make a GSI/LSI readable while the base table is unreadable. But in this series, we add the ability for an **LSI** (not GSI) query to use `Select=ALL_ATTRIBUTES` or `Select=SPECIFIC_ATTRIBUTES` and read from the base table. We need to add a base-table permissions check before we allow this. This extra check - and a test for it - is missing from this PR.

**Comment by nyh** [comment]
> I just realized there's something missing in this pull request - fixing RBAC (authorization):
> 
> As we test in `test_cql_rbac.py`::`test_rbac_query_separate_index_permissions`, a base table and its GSIs/LSIs have separate permissions - you can make a GSI/LSI readable while the base table is unreadable. But in this series, we add the ability for an **LSI** (not GSI) query to use `Select=ALL_ATTRIBUTES` or `Select=SPECIFIC_ATTRIBUTES` and read from the base table. We need to add a base-table permissions check before we allow this. This extra check - and a test for it - is missing from this PR.

This observation was outdated, it's not missing. We have a test `test_rbac_lsi_query_all_attributes_permissions`where we check that when reading with Select='ALL_ATTRIBUTES', we need SELECT permission on the base table, it's not enough to have permissions on the index table. This test is missing `Select=SPECIFIC_ATTRIBUTES` , I'll add it, but I don't think it's terribly important.

**Comment by nyh** [comment]
Rebased to fix conflicts, and added a test case for another case I mistakenly thought I didn't handle, but I did (but the test was missing anyway, so good to add it).

**Comment by nyh** [comment]
> #### Failed Tests (1/72095):
> 
>     * [test/cluster/tasks/test_tablet_tasks.py::test_tablet_repair_wait](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/30580/testReport/junit/cluster.tasks/test_tablet_tasks/debug___test_py___debug___test_tablet_repair_wait_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%22test_tablet_repair_wait.debug.2%22) [🎫](https://scylladb.atlassian.net/issues/?jql=text~%22test_tablet_repair_wait%22) [📂](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/30580/artifact/testlog/x86_64/debug/failed_test/test_tablet_repair_wait.debug.2/)

Known flaky test: https://scylladb.atlassian.net/browse/SCYLLADB-2531

**Comment by nyh** [comment]
Finally CI passed. @ScyllaPiotr please review (or ask someone to review).

**Comment by nyh** [comment]
> Finally CI passed. @ScyllaPiotr please review (or ask someone to review).

@ScyllaPiotr please review, this is a small feature, but one that users have asked for before (https://github.com/scylladb/scylladb/issues/5036) so it is worth considering.
